_drafts/practical_zephyr_01_kconfig.md

```diff
@@ -0,0 +1,986 @@
+
+
+### Adding a custom configuration
+
+[CMake automatically detects a `Kconfig` file](https://docs.zephyrproject.org/latest/develop/application/index.html#application-cmakelists-txt) if it is placed in the same directory of the application's `CMakeLists.txt`, and that is what we'll use for our own configuration file. In case you want to place the `Kconfig` file somewhere else, you can customize this behavior using an absolute path for the `KCONFIG_ROOT` build system variable.
+
+Similar to the file `zephyr/subsys/debug/Kconfig` of Zephyr's _debug_ subsystem, the `Kconfig` file specifies all available configuration options and their dependencies. We'll only specify a simple boolean symbol in the `Kconfig` file in our application's root directory. Please refer to the [official documentation](https://docs.zephyrproject.org/latest/build/kconfig/setting.html#setting-kconfig-configuration-values) for details about the `Kconfig` syntax:
+
+
+```bash
+$ tree --charset=utf-8 --dirsfirst -L 1
+.
+├── boards
+├── src
+├── CMakeLists.txt
+├── Kconfig
+├── extra0.conf
+├── extra1.conf
+├── prj.conf
+└── prj_release.conf
+
+$ cat Kconfig
+```
+```conf
+mainmenu "Customized Menu Name"
+source "Kconfig.zephyr"
+
+menu "Application Options"
+config USR_FUN
+	bool "Enable usr_fun"
+	default n
+	help
+	  Enables the usr_fun function.
+endmenu
+```
+
+The skeleton of this `Kconfig` file is well explained in the [official documentation](https://docs.zephyrproject.org/latest/develop/application/index.html#application-cmakelists-txt):
+
+The statement _"mainmenu"_ defines a custom text that is used as our `Kconfig` main menu. It is shown, e.g., as a title in the build target `menuconfig`. We'll see this in a moment when we'll build our target and launch `menuconfig`.
+
+The _"source"_ statement essentially includes the top-level Zephyr Kconfig file `zephyr/Kconfig.zephyr` and all of its symbols (all _"source"_ statements are relative to Zephyr's root directory). This is necessary since we're effectively replacing the `zephyr/Kconfig` file of the Zephyr base that is usually _parsed_ as the first file by _Kconfig_. We'll see this below when we look at the build output. The contents of the default root `Kconfig` file are quite similar to what we're doing right now:
+
+```bash
+$ cat $ZEPHYR_BASE/Kconfig
+# -- hidden comments --
+mainmenu "Zephyr Kernel Configuration"
+source "Kconfig.zephyr"
+```
+
+By sourcing the `Kconfig.zephr` file we're loading all _Kconfig_ menus and symbols provided with Zephr. Next, we declare our own menu between the _"menu"_ and _"endmenu"_ statements to group our application symbols. Within this menu, we declare our `USR_FUN` symbol, which we'll use to enable a function `usr_fun`.
+
+Let's rebuild our application without configuring `USR_FUN` and have a look at the build output:
+
+```bash
+$ west build --board nrf52840dk_nrf52840 -d ../build --pristine
```

Comment:
Maybe mention here that pristine is needed to pick up the new Kconfig file

Reply:
Repetition is key to remember things :) - thanks for pointing those out!